crates/apollo-compiler/src/validation/mod.rs

```diff
@@ -17,27 +17,56 @@ mod union_;
 mod value;
 mod variable;
 
+use crate::NodeStr;
 pub use validation_db::{ValidationDatabase, ValidationStorage};
 
+/// Track used names in a recursive function.
+struct RecursionStack {
+    seen: Vec<NodeStr>,
+}
+
+impl RecursionStack {
+    fn new() -> Self {
+        Self {
+            seen: Default::default(),
+        }
+    }
+
+    fn with_root(root: NodeStr) -> Self {
+        Self { seen: vec![root] }
+    }
+
+    /// Return the actual API for tracking recursive uses.
+    pub fn guard(&mut self) -> RecursionGuard<'_> {
+        RecursionGuard(&mut self.seen)
+    }
+}
+
 /// Track used names in a recursive function.
 ///
-/// Pass the result of `stack.push(name)` to recursive calls. Use `stack.contains(name)` to check
-/// if the name was used somewhere up the call stack.
-struct RecursionStack<'a>(&'a mut Vec<String>);
-impl RecursionStack<'_> {
-    fn push(&mut self, name: String) -> RecursionStack<'_> {
-        self.0.push(name);
-        RecursionStack(self.0)
+/// Pass the result of `guard.push(name)` to recursive calls. Use `guard.contains(name)` to check
+/// if the name was used somewhere up the call stack. When a guard is dropped, its name is removed
+/// from the list.
+struct RecursionGuard<'a>(&'a mut Vec<NodeStr>);
+impl RecursionGuard<'_> {
+    /// Mark that we saw a name.
+    fn push(&mut self, name: &NodeStr) -> RecursionGuard<'_> {
+        self.0.push(name.clone());
+        RecursionGuard(self.0)
     }
+    /// Check if we saw a name somewhere up the call stack.
     fn contains(&self, name: &str) -> bool {
         self.0.iter().any(|seen| seen == name)
```

Comment:
Could the `Vec` be replaced with `IndexSet` to make this O(1)?

Reply:
Huh, when I try that the structure seems to not work anymore, which is a bit suspicious. Looking into why that might be...

Reply:
It was suspicious because the directives recursion implementation was registering directive definitions twice!